### DISCUSSION TITLE
Should Trace Propagators define ConfigurablePropagatorProviders?

### DISCUSSION THREAD

**Discussion by kenfinnigan** [open-telemetry/opentelemetry-java#3284] [Q&A]
Currently, the [AWS propagator](https://github.com/open-telemetry/opentelemetry-java/tree/main/extensions/aws) implements `ConfigurablePropagatorProvider` from the AutoConfigure SPI, providing a convenient way for a propagator instance to be retrieved.

Is there any reason why the propagators in https://github.com/open-telemetry/opentelemetry-java/tree/main/extensions/trace-propagators don't define a `ConfigurablePropagatorProvider` for Jaeger, OTTracer, and B3?

As B3 has two types, it might require two providers. Beyond this, I'm not sure if there's a particular reason they're implemented differently?

**Comment by jkwatson** [comment]
Mostly, we don't want our stable propagators to depend on unstable auto-configuration code.

**Reply by kenfinnigan** [reply]
Which pieces of auto-configuration are unstable?

The whole module or just the bits that implement the SPI?

I'm wondering whether the SPI could be extracted and therefore be "stable" if the instability is not the SPI itself?

**Reply by jkwatson** [reply]
Well, the module as a whole hasn't been marked as non-alpha yet, so... all of it?

**Reply by kenfinnigan** [reply]
Fair enough.

Once it's non-alpha, would the above be considered? Or not really to still keep it simple?

**Reply by jkwatson** [reply]
I think we'd need a good use-case, since you can fully configure them via env vars/sysprops at the moment.

**Reply by kenfinnigan** [reply]
Ok thanks

**Reply by kenfinnigan** [reply]
Is there a way to mark this "resolved"? I couldn't find anything obvious

**Reply by jkwatson** [reply]
no idea. I think just adding some kind of +1 to your preferred answer ala stackoverflow?

**Reply by anuraaga** [reply]
I think it's ok to depend on the autoconfigure package from the propagators since it'd be `compileOnly`.

I did actually look at adding them but didn't see a need - autoconfigure already deals with them natively so it'd just be a different way of doing the same thing (it doesn't deal with aws natively since that's vendor specific and so does use the SPI).

@kenfinnigan do you have a use case for using the SPI for our standard propagators?

**Reply by kenfinnigan** [reply]
@anuraaga I was looking at implementing the ability to configure the propagators for Quarkus without necessarily having a classpath dependency on the specific implementations.

For example, being able to depend on the AutoConfigure SPI for finding and retrieving an `INSTANCE` for a propagator, but then what's in the classpath is determined by the users' application.

**Reply by anuraaga** [reply]
Thanks - I sent #3294 since it seems fine to me. Should we extract `autoconfigure-spi` into a separate artifact by the way?

**Reply by kenfinnigan** [reply]
@anuraaga If the SPI was separate that makes things simpler, as you could depend on the SPI without any of the actual autoconfigure code itself

**Reply by anuraaga** [reply]
Looked into extracting out the SPI - had a problem in that the SPI depends on `ConfigProperties` and `ConfigurationException` class. This seems to have two problems that

1) We'd need to include these non-SPI classes in the SPI. A bit weird
2) A different configuration mechanism like Quarkus presumably doesn't want to use `ConfigProperties`

@kenfinnigan Any thoughts on a good way to separate out the SPI?

**Reply by kenfinnigan** [reply]
@anuraaga I took a look at the code and I see there are a couple options:

1. Have `ConfigProperties` and `ConfigurationException` be a part of the SPI, as they're "required" for the SPI to work
2. Create an interface for `ConfigProperties`, which is part of the SPI, and the other SPI interfaces use, and then `ConfigProperties` implements it. This approach means `ConfigurationException` is only utilized within `ConfigProperties` and is not needed in the SPI
